### PR TITLE
Simplify fk::vector::resize(int)

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

Simplify `fk::vector::resize(int)` using new functions in asgard_resources.

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

File: src/asgard_vector.hpp [hunk 1]
```diff
@@ -1021,35 +1021,14 @@ fk::vector<P, mem, resrc>::resize(int const new_size)
...
-    {
-      if (size() < new_size)
-        copy_on_device(data_, old_data, size());
-      else
-        copy_on_device(data_, old_data, new_size);
-    }
-    delete_device(old_data);
+    memcpy_1d<resrc, resrc>(data_, old_data, std::min(size(), new_size));
   }
+  delete_resource<resrc>(old_data);
```
Comment (mkstoyanov) [inline src/asgard_vector.hpp]: What happens here, if `old_data` is `nullptr`

Comment (quantumsteve) [reply src/asgard_vector.hpp]: `delete[] (nullptr)` and `cudaFree(nullptr)` should do nothing.

https://en.cppreference.com/w/cpp/memory/new/operator_delete
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1ga042655cbbf3408f01061652a075e094

File: src/asgard_vector.hpp [hunk 2]
```diff
@@ -1023,20 +1023,26 @@ fk::vector<P, mem, resrc>::resize(int const new_size)
     return *this;
   P *old_data{data_};
 
-  if constexpr (resrc == resource::host)
+  if (new_size > 0)
   {
     data_ = new P[new_size];
+    std::fill_n(data_, new_size, P{0});
```
Comment (mkstoyanov) [inline src/asgard_vector.hpp]: Doesn't this merge with the command above and lead to:
```
  data_ = new P[new_size]();
```

Comment (quantumsteve) [reply src/asgard_vector.hpp]: Yes, it does the same thing and generates the same MPI test failure...

File: src/asgard_vector.hpp [hunk 3]
```diff
@@ -1036,20 +1036,21 @@ fk::vector<P, mem, resrc>::resize(int const new_size)
       else
         std::memcpy(data_, old_data, new_size * sizeof(P));
     }
-    delete[] old_data;
+    else if (size() == 0 && new_size > 0)
+    {
+      std::fill_n(data_, new_size, P{0});
+    }
```
Comment (quantumsteve) [inline src/asgard_vector.hpp]: @ckendrick @stefan-schnake this _could_ explain the issue you're encountering with odd behavior after multiple timesteps. Unfortunatley, the fix is failing CI tests (not on my local machine...)

Comment (stefan-schnake) [reply src/asgard_vector.hpp]: The issue we were having was unrelated to this issue and fixed in #625

**Review by mkstoyanov — APPROVED** [review]

**Pull request merged**